### PR TITLE
[otp/bazel] Refactor OTP Bazel rules and add overlay support

### PR DESCRIPTION
**Pull request by milesdai**
This commit refactors how the json configuration files are generated and adds the ability to provide multiple HJSON files as overlays to the `otp_image` target.

Addresses part of #15337.

### CODE REVIEW

File: rules/otp.bzl
Comment (dmcardle) [inline rules/otp.bzl:110]: WDYT about getting rid of `add_cfg` and making `src` an `attr.label_list`? Conceptually, this `otp_image` rule layers each JSON image config onto the previous one.

Something like this:

```bazel
        "image_configs": attr.label_list(
            allow_empty = False,
            allow_files = [".json", ".hjson"],
            doc = "Sequence of image configs in Hjson format. Later configs may clobber values from earlier configs.",
        ),
```

File: hw/ip/otp_ctrl/data/BUILD
Comment (dmcardle) [inline hw/ip/otp_ctrl/data/BUILD:286]: Nit: Move `:img_raw` and `:img_test_unlocked0` definitions to match the order of the `:otp_imgs` filegroup.

**Review by dmcardle — APPROVED** [review]
Lgtm! Thanks for working on this, Miles!

**Comment by dmcardle** [comment]
@milesdai and @cfrantz: WDYT about adding Bazel analysis tests for the I'm toying with the idea of adding some Bazel analysis tests for `otp_json` and `otp_image` rules?

File: rules/otp.bzl
Comment (milesdai) [reply rules/otp.bzl:74]: I have a docstring at the top of the file that explains the setup. I figured it would make more sense to explain the whole OTP generation flow in one place rather than splitting up the docs between several different functions.

File: rules/otp.bzl
Comment (milesdai) [reply rules/otp.bzl:110]: Yeah I considered consolidating all these inputs as `srcs`, but `gen-otp-img.py` treats the "base" configuration file slightly differently from the overlay files (i.e. those passed in via `--add_cfg`). For example, it requires that the base configuration have a `seed` parameter[^1].

I thought it would be better to explicitly have a separate parameter for the base configuration rather than doing something like "the first element of the `srcs` attribute will be used as the base configuration".

[^1]: Creating the image with the base `img_cfg`: https://cs.opensource.google/opentitan/opentitan/+/master:util/design/lib/OtpMemImg.py;l=150
Applying the overlay from `add_cfg`:
https://cs.opensource.google/opentitan/opentitan/+/master:util/design/lib/OtpMemImg.py;l=325

Comment (cfrantz) [reply rules/otp.bzl:110]: Since msf refers to the additional configs as layers, perhaps `layers` as the attribute name?

File: hw/ip/otp_ctrl/data/BUILD
Comment (cfrantz) [inline hw/ip/otp_ctrl/data/BUILD:24]: Are these translations of the existing hjson config files?  Are we planning on deleting those files at some point in favor of these in-bazel defined configs?

**Review by cfrantz — APPROVED** [review]

File: rules/otp.bzl
Comment (msfschaffner) [reply rules/otp.bzl:110]: Indeed - only the base config is allowed to have a seed value at the moment.
I believe this was done due to the way these additional config overlays are currently processed in the Python script.

I.e., the initial config is used to create the OTP image object and that involves seeding the RNG.
Additional overlays are then processed later on as "patches" to the base image - so changing the seed at a later point would interfere with how random constants are currently generated. We could loosen this constraint, but that would require a bit of refactoring so that the Python script first validates and combines all Hjson datastructures before creating an OTP image object (in hindsight this may have been a more efficient way of doing it).

File: hw/ip/otp_ctrl/data/BUILD
Comment (msfschaffner) [reply hw/ip/otp_ctrl/data/BUILD:24]: @moidx are we ok with moving all these definitions into Bazel rules, or do we want to keep the source of truth in separate Hjson files?

File: rules/otp.bzl
Comment (msfschaffner) [reply rules/otp.bzl:110]: > Since msf refers to the additional configs as layers, perhaps layers as the attribute name?

Or maybe `overlays`?

File: hw/ip/otp_ctrl/data/BUILD
Comment (moidx) [reply hw/ip/otp_ctrl/data/BUILD:24]: +1 to having a single source of truth in the long term so that it is easier to keep test cases in sync across targets (dv, fpga, verilator). 

Moving things to Bazel has the benefit of being able to customize OTP images as part of test target definitions. I believe only ROM E2E tests running on FPGA/verilator are taking advantage of this. I assume that once we start running those tests in DV we will need to add support for this in dvsim.

@msfschaffner can you outline what we would be missing if we move these configs to Bazel? One thing I think it would be good to expose is the seed value as a DV configurable param. Are there any other things we are missing?

Here is some additional context on the DV side:

## Current dv configuration

In the current DV target configuration, OTP images need to be predefined in the dvsim configuration. For example: 

1. `chip_sim_cfg.hjson`: pre-run commands, e.g.

```hjson
{gen_otp_images_cmd} \
              --img-cfg {gen_otp_images_cfg_dir}/otp_ctrl_img_dev.hjson \
              --add-cfg {gen_otp_images_cfg_dir}/otp_ctrl_img_creator_sw_cfg.hjson \
              --add-cfg {gen_otp_images_cfg_dir}/otp_ctrl_img_owner_sw_cfg.hjson \
              --add-cfg {gen_otp_images_cfg_dir}/otp_ctrl_img_hw_cfg.hjson \
              --out {run_dir}/otp_ctrl_img_dev.vmem \
              {gen_otp_images_cmd_opts}
```

4. `hw/top_earlgrey/dv/env/chip_env_cfg.sv`: Mapping to DV UVM params, e.g.

```sv
    otp_images[lc_ctrl_state_pkg::LcStRaw] = "otp_ctrl_img_raw.vmem";
    otp_images[lc_ctrl_state_pkg::LcStDev] = "otp_ctrl_img_dev.vmem";
    otp_images[lc_ctrl_state_pkg::LcStProd] = "otp_ctrl_img_prod.vmem";
    otp_images[lc_ctrl_state_pkg::LcStRma] = "otp_ctrl_img_rma.vmem";
    otp_images[lc_ctrl_state_pkg::LcStTestUnlocked0] = "otp_ctrl_img_test_unlocked0.vmem";
```

5. `chip_sim_cfg.hjson`: Request specific OTP configuration via parameter setting, e.g. `+use_otp_image=LcStTestUnlocked0`

```
    {
      name: chip_sw_csrng_lc_hw_debug_en_test
      uvm_test_seq: chip_sw_csrng_lc_hw_debug_en_vseq
      sw_images: ["//sw/device/tests/sim_dv:csrng_lc_hw_debug_en_test:1"]
      en_run_modes: ["sw_test_mode_test_rom"]
      run_opts: ["+sw_test_timeout_ns=15_000_000", "+rng_srate_value_min=15",
                 "+use_otp_image=LcStTestUnlocked0"]
      run_timeout_mins: 60
    }
```
Note: DV still has the ability to inject OTP values via backdoor, so there is still a chance values will not match across targets. The backdoor is required to optimize simulation time (otherwise it would take took long to run the same tests with the foundry database).

## Moving to Bazel OTP targets

The following could be done to start aligning DV with FPGA build infra. We would have to make changes to DV infra to be able to consume the OTP image target.

1. `chip_sim_cfg.hjson`: Request specific OTP configuration via build target:

```
    {
      name: chip_sw_csrng_lc_hw_debug_en_test
      uvm_test_seq: chip_sw_csrng_lc_hw_debug_en_vseq
      sw_images: ["//sw/device/tests/sim_dv:csrng_lc_hw_debug_en_test:1"]
      otp_images: ["//hw/ip/otp_ctrl/data:otp_test_unlocked0_image"]
      en_run_modes: ["sw_test_mode_test_rom"]
      run_opts: ["+sw_test_timeout_ns=15_000_000", "+rng_srate_value_min=15"]
      run_timeout_mins: 60
    }
```

Note:  Currently, the colon after the sw_image target name (e.g. :1) indicates the image type: 

```sv
  typedef enum {
    SwTypeRom,  // Ibex SW - first stage boot ROM.
    SwTypeTest, // Ibex SW - actual test SW.
    SwTypeOtbn  // Otbn SW.
  } sw_type_e;
```

Adding @sriyerg  since this is getting into DV land. 

If we need to discuss this in more detail, perhaps it can be added to the next top-level tests agenda?

CC @timothytrippel

Comment (sriyerg) [reply hw/ip/otp_ctrl/data/BUILD:24]: I am not aware of the background for this change. Could you please share the doc? I would expect the way to handle this would be to simply replace the python invocations with bazel invocations, and somehow retain the ability to pass the seed value to the invocation.

File: rules/otp.bzl
Comment (milesdai) [reply rules/otp.bzl]: Changed the attr name to `overlays`

File: rules/otp.bzl
Comment (msfschaffner) [inline rules/otp.bzl:49]: A quick heads-up about seeds: 

Right now, the DV environment will also need the ability to optionally pass the `--lc-seed` and `--otp-seed` arguments to the generator script. 

I should note though that after M2 we will have to replace the RNG mechanism used in these scripts by the entropy buffer file mechanism explained in https://github.com/lowRISC/opentitan/issues/15144 (see doc linked on that issue). 

For simulation / emulation purposes, reseeding the design and OTP image then essentially becomes a two-step process where a separate script will have to be called to generate new buffer files prior to the SW and HW builds (topgen will have to be rerun as well). For the tapeout on the other hand, the plan is to use an HSM in the secure domain to create high quality entropy buffer files.

The seed arguments should then likely become file paths pointing to individual entropy buffer files (they can be handled by BAZEL dependency management as well). We can discuss this more once this initial PR has landed, though.

**Comment by milesdai** [comment]
From our discussion over the past couple weeks, I believe this PR should be good to go. There will need to be additional modifications to how the seed is passed into the rule, and additional work will be needed to integrate this into the DV tools, but this PR should not impact that. As such, are we good to merge this one @msfschaffner?

**Comment by alphan** [comment]
Merged this unblock several other PRs. We can improve on this once we reach a consensus in #15337.

**Comment by msfschaffner** [comment]
Sounds good to me!

**Pull request merged**
